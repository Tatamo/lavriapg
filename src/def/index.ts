export * from "./language";
export * from "./parsingtable";
export * from "./token";
