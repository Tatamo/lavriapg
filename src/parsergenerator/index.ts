export * from "./closureset";
export * from "./closureitem";
export * from "./dfagenerator";
export * from "./firstset";
export * from "./nullableset";
export * from "./parsergenerator";
export * from "./symboldiscriminator";
export * from "./syntaxdb";
