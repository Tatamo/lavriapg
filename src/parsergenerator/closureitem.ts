import {Token} from "../def/token";
import {SyntaxDB} from "./syntaxdb";

export class ClosureItem {
	// インスタンス生成後に内部状態が変化することはないものとする
	private _lr0_hash: string;
	private _lr1_hash: string;
	constructor(private syntax: SyntaxDB, private _syntax_id: number, private _dot_index: number, private _lookaheads: Array<Token>) {
		// 有効な値かどうか調べる
		if (!this.syntax.hasDefinitionId(this._syntax_id)) {
			throw new Error("invalid syntax id");
		}
		if (this._dot_index < 0 || this._dot_index > this.syntax.getDefinitionById(this._syntax_id).pattern.length) {
			throw new Error("dot index out of range");
		}
		if (this._lookaheads.length == 0) {
			// 必要か？
			throw new Error("one or more lookahead symbols needed");
		}
		this.sortLA();
		this.updateHash();
	}
	get syntax_id(): number {
		return this._syntax_id;
	}
	get dot_index(): number {
		return this._dot_index;
	}
	get lookaheads(): Array<Token> {
		return this._lookaheads;
	}
	private sortLA() {
		this.lookaheads.sort((t1: Token, t2: Token) => {
			return this.syntax.getTokenId(t1) - this.syntax.getTokenId(t2);
		});
	}
	// ハッシュ文字列を生成する
	private updateHash() {
		this._lr0_hash = this.syntax_id.toString() + "," + this.dot_index.toString();
		let la_hash = "[";
		for (let i = 0; i < this.lookaheads.length; i++) {
			la_hash += this.syntax.getTokenId(this.lookaheads[i]).toString();
			if (i != this.lookaheads.length - 1) la_hash += ",";
		}
		la_hash += "]";
		this._lr1_hash = this._lr0_hash + "," + la_hash;
	}
	// 先読み部分を除いたハッシュ文字列を取得
	public getLR0Hash(): string {
		return this._lr0_hash;
	}
	// 先読み部分を含めたハッシュ文字列を取得
	public getLR1Hash(): string {
		return this._lr1_hash;
	}
	// 先読み部分を除いた部分が一致しているか調べる
	public isSameLR0(c: ClosureItem): boolean {
		return this.getLR0Hash() == c.getLR0Hash();
	}
	// 先読み部分も含めて完全に一致しているか調べる
	public isSameLR1(c: ClosureItem): boolean {
		return this.getLR1Hash() == c.getLR1Hash();
	}
	// LR0部分の同じ2つのClosureItemの先読み部分を統合して新しいClosureItemを生成する
	public merge(c: ClosureItem): ClosureItem | null {
		// LR0部分が違っている場合はnullを返す
		if (!this.isSameLR0(c)) return null;
		// LR1部分まで同じ場合は自身を返す
		if (this.isSameLR1(c)) return this;
		// 双方のlookaheads配列はソート済みであるとする
		let i1 = 0;
		let i2 = 0;
		const new_la = [];
		// 2つのLA配列をマージして新しい配列を生成する
		while (i1 < this.lookaheads.length || i2 < c.lookaheads.length) {
			if (i1 == this.lookaheads.length) {
				new_la.push(c.lookaheads[i2++]);
			}
			else if (i2 == c.lookaheads.length) {
				new_la.push(this.lookaheads[i1++]);
			}
			else if (this.lookaheads[i1] == c.lookaheads[i2]) {
				new_la.push(this.lookaheads[i1++]);
				i2++;
			}
			else if (this.syntax.getTokenId(this.lookaheads[i1]) < this.syntax.getTokenId(c.lookaheads[i2])) {
				new_la.push(this.lookaheads[i1++]);
			}
			else {
				new_la.push(c.lookaheads[i2++]);
			}
		}
		return new ClosureItem(this.syntax, this.syntax_id, this.dot_index, new_la);
	}
}
