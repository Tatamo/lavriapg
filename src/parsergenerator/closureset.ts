import {Token} from "../def/token";
import {ClosureItem} from "./closureitem";
import {GrammarDB} from "./grammardb";

/**
 * 複数のLRアイテムを保持するアイテム集合であり、インスタンス生成時に自身をクロージャー展開する
 *
 * [[GrammarDB]]から与えられるトークンIDをもとにして、LR(0)およびLR(1)アイテム集合としてのハッシュ値を生成することができる
 *
 * Immutableであるべきオブジェクトであるため、インスタンス生成後は内部状態が変化することはないと仮定される
 */
export class ClosureSet {
	// インスタンス生成後に内部状態が変化することはないものとする
	private _lr0_hash: string;
	private _lr1_hash: string;
	/**
	 * @param {GrammarDB} grammardb 使用する構文の情報
	 * @param {Array<ClosureItem>} closureset
	 */
	constructor(private grammardb: GrammarDB, private closureset: Array<ClosureItem>) {
		this.expandClosure();
		this.sort();
		this.updateHash();
	}
	/**
	 * 自身が保持する複数の[[ClosureItem]]は、常にLR(1)ハッシュによってソートされた状態に保たれているようにする
	 */
	private sort() {
		this.closureset.sort((i1: ClosureItem, i2: ClosureItem) => {
			if (i1.getLR1Hash() < i2.getLR1Hash()) return -1;
			else if (i1.getLR1Hash() > i2.getLR1Hash()) return 1;
			return 0;
		});
	}
	/**
	 * 保持しているLRアイテムの数
	 */
	get size() {
		return this.closureset.length;
	}
	/**
	 * 保持している[[ClosureItem]]の配列を得る
	 * @param {boolean} prevent_copy trueを与えると配列をコピーせず返す
	 *
	 * 得られた配列に変更が加えられないと保証される場合に用いる
	 * @returns {Array<ClosureItem>}
	 */
	public getArray(prevent_copy: boolean = false): Array<ClosureItem> {
		if (prevent_copy) return this.closureset;
		// デフォルトではコピーして返す(パフォーマンスは少し落ちる)
		return this.closureset.concat();
	}
	/**
	 * LRアイテムが集合に含まれているかどうかを調べる
	 *
	 * @param {ClosureItem} item
	 * @returns {boolean}
	 */
	public includes(item: ClosureItem): boolean {
		// 二分探索を用いて高速に探索する
		let min = 0;
		let max = this.closureset.length - 1;
		while (min <= max) {
			const mid = min + Math.floor((max - min) / 2);
			if (item.getLR1Hash() < this.closureset[mid].getLR1Hash()) {
				max = mid - 1;
			}
			else if (item.getLR1Hash() > this.closureset[mid].getLR1Hash()) {
				min = mid + 1;
			}
			else {
				// itemとclosureset[mid]が等しい
				return true;
			}
		}
		return false;
	}
	/**
	 * LR(0)ハッシュの一致を調べる
	 * @param {ClosureSet} cs 比較対象のアイテム集合
	 * @returns {boolean}
	 */
	public isSameLR0(cs: ClosureSet): boolean {
		return this.getLR0Hash() == cs.getLR0Hash();
	}
	/**
	 * LR(1)ハッシュの一致を調べる
	 * @param {ClosureSet} cs 比較対象のアイテム集合
	 * @returns {boolean}
	 */
	public isSameLR1(cs: ClosureSet): boolean {
		return this.getLR1Hash() == cs.getLR1Hash();
	}
	/**
	 * ハッシュ文字列を生成する
	 */
	private updateHash() {
		let lr0_hash = "";
		let lr1_hash = "";
		for (let i = 0; i < this.closureset.length; i++) {
			lr0_hash += this.closureset[i].getLR0Hash();
			lr1_hash += this.closureset[i].getLR1Hash();
			if (i != this.closureset.length - 1) {
				lr0_hash += "|";
				lr1_hash += "|";
			}
		}
		this._lr0_hash = lr0_hash;
		this._lr1_hash = lr1_hash;
	}
	/**
	 * LR(0)アイテム集合としてのハッシュ文字列を得る
	 * @returns {string}
	 */
	public getLR0Hash() {
		return this._lr0_hash;
	}
	/**
	 * LR(1)アイテム集合としてのハッシュ文字列を得る
	 * @returns {string}
	 */
	public getLR1Hash() {
		return this._lr1_hash;
	}
	/**
	 * LR(0)部分が同じ2つのClosureSetについて、先読み部分を統合した新しいClosureSetを生成する
	 *
	 * 異なるLR(0)アイテム集合であった場合、nullを返す
	 * @param {ClosureSet} cs マージ対象のアイテム集合
	 * @returns {ClosureSet | null} 先読み部分がマージされた新しいアイテム集合
	 */
	public mergeLA(cs: ClosureSet): ClosureSet | null {
		// LR0部分が違っている場合はnullを返す
		if (!this.isSameLR0(cs)) return null;
		// LR1部分まで同じ場合は自身を返す
		if (this.isSameLR1(cs)) return this;
		const a1 = this.getArray();
		const a2 = cs.getArray();
		const new_set: Array<ClosureItem> = [];
		// 2つの配列においてLR部分は順序を含めて等しい
		for (let i = 0; i < a1.length; i++) {
			const new_item = a1[i].merge(a2[i]);
			if (new_item != null) new_set.push(new_item);
		}
		return new ClosureSet(this.grammardb, new_set);
	}

	/**
	 * クロージャー展開を行う
	 *
	 * TODO: リファクタリング
	 */
	private expandClosure() {
		// 展開処理中はClosureItemのlookaheadsの要素数を常に1に保つこととする
		// 初期化
		const set: Array<ClosureItem> = [];
		// ClosureItemをlookaheadsごとに分解する
		for (const ci of this.closureset) {
			for (const la of ci.lookaheads) {
				set.push(new ClosureItem(this.grammardb, ci.rule_id, ci.dot_index, [la]));
			}
		}
		this.closureset = set;
		this.sort();

		// 変更がなくなるまで繰り返す
		let index = 0;
		while (index < this.closureset.length) {
			const ci = this.closureset[index++];
			const pattern = this.grammardb.getRuleById(ci.rule_id).pattern;

			if (ci.dot_index == pattern.length) continue; // .が末尾にある場合はスキップ
			const follow = pattern[ci.dot_index];
			if (!this.grammardb.symbols.isNonterminalSymbol(follow)) continue; // .の次の記号が非終端記号でないならばスキップ

			// クロージャー展開を行う

			// 先読み記号を導出
			// ci.lookaheadsは要素数1のため、0番目のインデックスのみを参照すればよい
			const lookaheads = [...this.grammardb.first.get(pattern.slice(ci.dot_index + 1).concat(ci.lookaheads[0])).values()];
			lookaheads.sort((t1: Token, t2: Token) => {
				return this.grammardb.getTokenId(t1) - this.grammardb.getTokenId(t2);
			});

			// symbolを左辺にもつ全ての規則を、先読み記号を付与して追加
			const rules = this.grammardb.findRules(follow);
			for (const {id} of rules) {
				for (const la of lookaheads) {
					const new_ci = new ClosureItem(this.grammardb, id, 0, [la]);
					// 重複がなければ新しいアイテムを追加する
					let flg_duplicated = false;
					for (const existing_item of this.closureset) {
						if (new_ci.isSameLR1(existing_item)) {
							flg_duplicated = true;
							break;
						}
					}
					if (!flg_duplicated) {
						this.closureset.push(new_ci);
					}
				}
			}
		}
		this.sort();

		// ClosureItemの先読み部分をマージする
		const tmp = this.closureset;
		this.closureset = [];
		let merged_lookaheads = [];
		for (let i = 0; i < tmp.length; i++) {
			merged_lookaheads.push(tmp[i].lookaheads[0]);
			if (i == tmp.length - 1 || !tmp[i].isSameLR0(tmp[i + 1])) {
				this.closureset.push(new ClosureItem(this.grammardb, tmp[i].rule_id, tmp[i].dot_index, merged_lookaheads));
				merged_lookaheads = [];
			}
		}
	}
}
