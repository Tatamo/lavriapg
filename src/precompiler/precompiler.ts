import {Language} from "../def/language";
import {SYMBOL_EOF} from "../def/token";
import {ParserGenerator} from "../parsergenerator/parsergenerator";
import {language_parser} from "./ruleparser";

/**
 * 予め構文解析器を生成しておいて利用するためのソースコードを生成する
 */
export class PreCompiler {
	/**
	 * @param import_path パーサジェネレータをimportするためのディレクトリパス
	 */
	constructor(private import_path: string = "lavriapg") {
		if (import_path[import_path.length - 1] != "/") this.import_path += "/";
	}
	/**
	 * 構文ファイルを受け取り、それを処理できるパーサを構築するためのソースコードを返す
	 * @param {string} input 言語定義文法によって記述された、解析対象となる言語
	 * @returns {string} 生成されたパーサのソースコード
	 */
	public exec(input: string): string {
		const language: Language = language_parser.parse(input);
		console.error(language);
		const parsing_table = new ParserGenerator(language).getParsingTable();
		let result = "";

		result += 'import {Token, SYMBOL_EOF} from "' + this.import_path + 'def/token";\n';
		result += 'import {Language} from "' + this.import_path + 'def/language";\n';
		result += 'import {ParsingOperation, ParsingTable} from "' + this.import_path + 'def/parsingtable";\n';
		result += 'import {Parser} from "' + this.import_path + 'parser/parser";\n';
		result += 'import {ParserFactory} from "' + this.import_path + 'parser/factory";\n\n';

		/*
		result += 'import {Token, SYMBOL_EOF, Language, ParsingOperation, ParsingTable, Parser, ParserFactory} from "' + this.import_path +'";\n\n';
	   */

		result += "export const language: Language = {\n";
		result += "\t" + "lex: {rules:[\n";
		for (let i = 0; i < language.lex.rules.length; i++) {
			const token = language.lex.rules[i].token;
			const pattern = language.lex.rules[i].pattern;
			result += "\t\t" + "{token: " + (token === null ? "null" : ('"' + (token as string)) + '"') + ", pattern: ";
			if (pattern instanceof RegExp) {
				result += pattern.toString();
			}
			else {
				result += '"' + pattern + '"';
			}
			result += "}";
			if (i != language.lex.rules.length - 1) result += ",";
			result += "\n";
		}
		result += "\t" + "]},\n";
		result += "\t" + "grammar: {rules:[\n";
		for (let i = 0; i < language.grammar.rules.length; i++) {
			const ltoken = language.grammar.rules[i].ltoken;
			const pattern = language.grammar.rules[i].pattern;
			result += "\t\t" + "{\n";
			result += "\t\t\t" + 'ltoken: "' + (ltoken as string) + '",\n';
			result += "\t\t\t" + "pattern: [";
			for (let ii = 0; ii < pattern.length; ii++) {
				result += '"' + (pattern[ii] as string) + '"';
				if (ii != pattern.length - 1) result += ", ";
			}
			result += "]\n";
			result += "\t\t" + "}";
			if (i != language.grammar.rules.length - 1) result += ",";
			result += "\n";
		}
		result += "\t" + "],\n";
		result += "\t" + 'start_symbol: "' + (language.grammar.start_symbol as string) + '"\n';
		result += "\t" + "},\n";
		result += "};\n\n";
		result += "export const parsing_table:ParsingTable = [\n";
		for (let i = 0; i < parsing_table.length; i++) {
			result += "\t" + "new Map<Token, ParsingOperation>([\n";
			parsing_table[i].forEach((value, key) => {
				result += "\t\t" + "[" + (key == SYMBOL_EOF ? "SYMBOL_EOF" : ('"' + (key as string)) + '"') + ", " + JSON.stringify(value) + "],\n";
			});
			result = result.slice(0, -2);
			result += " ]),\n";
		}
		result = result.slice(0, -2);
		result += "\n];\n\n";
		result += "export const parser:Parser = ParserFactory.create(language, parsing_table);\n";
		return result;
	}
}
