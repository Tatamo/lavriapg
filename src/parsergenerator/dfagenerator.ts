import {SYMBOL_EOF, Token} from "../def/token";
import {ClosureItem} from "./closureitem";
import {ClosureSet} from "./closureset";
import {SyntaxDB} from "./syntaxdb";

export type DFAEdge = Map<Token, number>;
export type DFANode = { closure: ClosureSet, edge: DFAEdge };
export type DFA = Array<DFANode>;

export class DFAGenerator {
	private lr_dfa: DFA;
	private lalr_dfa: DFA;
	constructor(private syntax: SyntaxDB) {
		this.generateDFA();
	}
	public getLR1DFA(): DFA {
		return this.lr_dfa;
	}
	public getLALR1DFA(): DFA {
		return this.lalr_dfa;
	}
	// DFAの生成
	private generateDFA() {
		const initial_item: ClosureItem = new ClosureItem(this.syntax, -1, 0, [SYMBOL_EOF]);
		const initial_set: ClosureSet = new ClosureSet(this.syntax, [initial_item]);
		const dfa: DFA = [{closure: initial_set, edge: new Map<Token, number>()}];

		// 変更がなくなるまでループ
		let flg_changed = true;
		let i = 0;
		while (flg_changed) {
			flg_changed = false;
			while (i < dfa.length) {
				const closure = dfa[i].closure;
				const edge = dfa[i].edge;
				const new_sets: Map<Token, ClosureSet> = this.generateNewClosureSets(closure);

				// 新しいノードを生成する
				for (const [edge_label, cs] of new_sets) {
					const new_node: DFANode = {closure: cs, edge: new Map<Token, number>()};
					// 既存のNodeのなかに同一のClosureSetを持つものがないか調べる
					const duplicated_index = this.indexOfDuplicatedNode(dfa, new_node);
					let index_to;
					if (duplicated_index == -1) {
						// 既存の状態と重複しない
						dfa.push(new_node);
						index_to = dfa.length - 1;
						flg_changed = true;
					}
					else {
						// 既存の状態と規則が重複する
						// 新しいノードの追加は行わず、重複する既存ノードに対して辺を張る
						index_to = duplicated_index;
					}
					// 辺を追加する
					if (!edge.has(edge_label)) {
						edge.set(edge_label, index_to);
						// 新しい辺が追加された
						flg_changed = true;
						// DFAを更新
						dfa[i] = {closure, edge};
					}
				}
				i++;
			}
			i = 0;
		}
		this.lr_dfa = dfa;
		this.lalr_dfa = this.mergeLA(dfa);
	}
	// TODO: バグがないか確認
	// LR(1)オートマトンの先読み部分をマージして、LALR(1)オートマトンを作る
	private mergeLA(dfa: DFA): DFA {
		const array: Array<DFANode | null> = dfa.slice(); // nullを許容する
		const merge_to: Map<number, number> = new Map<number, number>(); // マージ先への対応関係を保持する

		for (let i = 0; i < array.length; i++) {
			if (array[i] == null) continue;
			for (let ii = i + 1; ii < array.length; ii++) {
				if (array[ii] == null) continue;
				// LR(0)アイテムセット部分が重複
				if (array[i]!.closure.isSameLR0(array[ii]!.closure)) {
					// ii番目の先読み部分をi番目にマージする
					// インデックス番号の大きい方が削除される
					// 辺情報は、対象となる辺もいずれマージされて消えるため操作しなくてよい

					// 更新
					// Nodeに変更をかけるとLR(1)DFAの中身まで変化してしまうため新しいオブジェクトとする
					array[i] = {closure: array[i]!.closure.mergeLA(array[ii]!.closure)!, edge: array[i]!.edge};
					// ii番目を削除
					array[ii] = null;
					// マージ元->マージ先への対応関係を保持
					merge_to.set(ii, i);
				}
			}
		}
		// 削除した部分を配列から抜き取る
		const prev_length = array.length; // ノードをマージする前のノード総数
		const fix = new Array(prev_length); // (元のindex->現在のindex)の対応表を作る
		let d = 0; // ずれ
		// nullで埋めた部分を消すことによるindexの変化
		for (let i = 0; i < prev_length; i++) {
			if (array[i] === null) d += 1; // ノードが削除されていた場合、以降のインデックスを1つずらす
			else fix[i] = i - d;
		}
		// 配列からnull埋めした部分を削除したものを作る
		const shortened: Array<DFANode> = [];
		for (const node of array) {
			if (node !== null) shortened.push(node);
		}
		// fixのノードが削除された部分を埋める
		for (const [from, to] of merge_to) {
			let index = to;
			while (merge_to.has(index)) index = merge_to.get(index)!;
			fix[from] = fix[index]; // toを繰り返し辿っているので未定義部分へのアクセスは発生しない
		}

		const result: DFA = [];
		// インデックスの対応表をもとに辺情報を書き換える
		for (const node of shortened) {
			const new_edge = new Map<Token, number>();
			for (const [token, node_index] of node.edge) {
				new_edge.set(token, fix[node_index]);
			}
			result.push({closure: node.closure, edge: new_edge});
		}
		return result;
	}
	// 既存のClosureSetから新しい規則を生成し、対応する記号ごとにまとめる
	private generateNewClosureSets(closureset: ClosureSet): Map<Token, ClosureSet> {
		const tmp: Map<Token, Array<ClosureItem>> = new Map<Token, Array<ClosureItem>>();
		// 規則から新しい規則を生成し、対応する記号ごとにまとめる
		for (const {syntax_id, dot_index, lookaheads} of closureset.getArray()) {
			const pattern = this.syntax.getDefinitionById(syntax_id).pattern;
			if (dot_index == pattern.length) continue; // .が末尾にある場合はスキップ
			const new_ci = new ClosureItem(this.syntax, syntax_id, dot_index + 1, lookaheads);
			const edge_label: Token = pattern[dot_index];

			let items: Array<ClosureItem>;
			if (tmp.has(edge_label)) {
				// 既に同じ記号が登録されている
				items = tmp.get(edge_label)!;
			}
			else {
				// 同じ記号が登録されていない
				items = [];
			}
			items.push(new_ci);
			tmp.set(edge_label, items);
		}
		// ClosureItemの配列からClosureSetに変換
		const result: Map<Token, ClosureSet> = new Map<Token, ClosureSet>();
		for (const [edge_label, items] of tmp) {
			result.set(edge_label, new ClosureSet(this.syntax, items));
		}
		return result;
	}
	// 与えられたDFANodeと全く同じDFANodeがある場合、そのindexを返す
	// 見つからなければ-1を返す
	private indexOfDuplicatedNode(dfa: DFA, new_node: DFANode): number {
		let index = -1;
		for (const [i, node] of dfa.entries()) {
			if (new_node.closure.isSameLR1(node.closure)) {
				index = i;
				break;
			}
		}
		return index;
	}
}
