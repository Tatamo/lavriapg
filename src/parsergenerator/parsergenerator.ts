import {Language} from "../def/language";
import {ParsingTable, AcceptOperation, ConflictedOperation, GotoOperation, ParsingOperation, ReduceOperation, ShiftOperation} from "../def/parsingtable";
import {SYMBOL_EOF, Token} from "../def/token";
import {ParserFactory} from "../parser/factory";
import {Parser} from "../parser/parser";
import {DFA, DFAGenerator} from "./dfagenerator";
import {GrammarDB} from "./grammardb";

export class ParserGenerator {
	private parsing_table: ParsingTable;
	private table_type: "LR1" | "LALR1" | "CONFLICTED";
	private grammardb: GrammarDB;
	private dfa_generator: DFAGenerator;

	constructor(private language: Language) {
		this.grammardb = new GrammarDB(this.language);
		this.dfa_generator = new DFAGenerator(this.grammardb);
		this.init();
	}

	private init() {
		const lalr_result = this.generateParsingTable(this.dfa_generator.getLALR1DFA());
		if (lalr_result.success) {
			this.parsing_table = lalr_result.table;
			this.table_type = "LALR1";
			return;
		}
		// LALR(1)構文解析表の生成に失敗
		// LR(1)構文解析表の生成を試みる
		console.error("LALR parsing conflict found. use LR(1) table.");
		const lr_result = this.generateParsingTable(this.dfa_generator.getLR1DFA());
		this.parsing_table = lr_result.table;
		this.table_type = "LR1";
		if (!lr_result.success) {
			// LR(1)構文解析表の生成に失敗
			this.table_type = "CONFLICTED";
			console.error("LR(1) parsing conflict found. use LR(1) conflicted table.");
		}
	}

	public getParser(): Parser {
		return ParserFactory.create(this.language, this.parsing_table);
	}

	public getParsingTable(): ParsingTable {
		return this.parsing_table;
	}

	public isConflicted(): boolean {
		return this.table_type === "CONFLICTED";
	}

	public getTableType(): "LR1" | "LALR1" | "CONFLICTED" {
		return this.table_type;
	}

	// 構文解析表を構築する
	private generateParsingTable(dfa: DFA) {
		const parsing_table: ParsingTable = [];
		let flg_conflicted = false;

		for (const node of dfa) {
			const table_row = new Map<Token, ParsingOperation>();
			// 辺をもとにshiftとgotoオペレーションを追加
			for (const [label, to] of node.edge) {
				if (this.grammardb.symbols.isTerminalSymbol(label)) {
					// ラベルが終端記号の場合
					// shiftオペレーションを追加
					const operation: ShiftOperation = {type: "shift", to};
					table_row.set(label, operation);
				}
				else if (this.grammardb.symbols.isNonterminalSymbol(label)) {
					// ラベルが非終端記号の場合
					// gotoオペレーションを追加
					const operation: GotoOperation = {type: "goto", to};
					table_row.set(label, operation);
				}
			}

			// Closureをもとにacceptとreduceオペレーションを追加していく
			for (const item of node.closure.getArray()) {
				// 規則末尾が.でないならスキップ
				// if(item.pattern.getRuleById(item.pattern.size-1) != SYMBOL_DOT) return;
				if (item.dot_index != this.grammardb.getRuleById(item.rule_id).pattern.length) continue;
				if (item.rule_id == -1) {
					// acceptオペレーション
					// この規則を読み終わると解析終了
					// $をラベルにacceptオペレーションを追加
					const operation: AcceptOperation = {type: "accept"};
					table_row.set(SYMBOL_EOF, operation);
					continue;
				}
				for (const label of item.lookaheads) {
					const operation: ReduceOperation = {type: "reduce", grammar_id: item.rule_id};
					// 既に同じ記号でオペレーションが登録されていないか確認

					if (table_row.has(label)) {
						// コンフリクトが発生
						flg_conflicted = true; // 構文解析に失敗
						const existing_operation = table_row.get(label)!; // 上で.has(label)のチェックを行っているためnon-nullable
						const conflicted_operation: ConflictedOperation = {type: "conflict", shift_to: [], reduce_grammar: []};
						if (existing_operation.type == "shift") {
							// shift/reduce コンフリクト
							conflicted_operation.shift_to = [existing_operation.to];
							conflicted_operation.reduce_grammar = [operation.grammar_id];
						}
						else if (existing_operation.type == "reduce") {
							// reduce/reduce コンフリクト
							conflicted_operation.shift_to = [];
							conflicted_operation.reduce_grammar = [existing_operation.grammar_id, operation.grammar_id];
						}
						else if (existing_operation.type == "conflict") {
							// もっとやばい衝突
							conflicted_operation.shift_to = existing_operation.shift_to;
							conflicted_operation.reduce_grammar = existing_operation.reduce_grammar.concat([operation.grammar_id]);
						}
						// とりあえず衝突したオペレーションを登録しておく
						table_row.set(label, conflicted_operation);
					}
					else {
						// 衝突しないのでreduceオペレーションを追加
						table_row.set(label, operation);
					}
				}
			}
			parsing_table.push(table_row);
		}
		// if (flg_conflicted) console.error("warn: some conflicts may be occured");
		return {table: parsing_table, success: !flg_conflicted};
	}
}
