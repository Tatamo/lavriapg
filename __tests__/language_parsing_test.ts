import {language_language, language_parser} from "../src/precompiler/ruleparser";
import {ParserGenerator} from "../src/parsergenerator/parsergenerator";
import {Language} from "../src";

describe("language parsing test", () => {
	const input = require("fs").readFileSync("language", "utf8");
	const removeCallback = (language: Language): Language => {
		const lex = {...language.lex};
		lex.rules = lex.rules.map(({token, pattern, states}) => ({token, pattern, states}));
		const grammar = {...language.grammar};
		grammar.rules = grammar.rules.map(({ltoken, pattern}) => ({ltoken, pattern}));
		return {lex, grammar};
	};

	const language_language_without_callback = removeCallback(language_language);
	const pg = new ParserGenerator(language_language);
	test("valid parser", () => {
		expect(pg.isConflicted()).toBeFalsy();
	});
	const parser = pg.getParser(); // language_parserと同一のものであることが期待される
	test("parsing language file", () => {
		expect(removeCallback(parser.parse(input))).toEqual(language_language_without_callback);
	});
	// languageファイルを読み取ってパーサを生成したい
	test("language_parser", () => {
		expect(removeCallback(language_parser.parse(input))).toEqual(language_language_without_callback);
	});
});

describe("syntax functions test", () => {
	const pg = new ParserGenerator(language_language);
	const parser = pg.getParser();
	test("lex-state", () => {
		const input = `
A	"a"
<state1, state2>B	"b"
<default>B2	"b"
C	"c"
$S : A B2 C;
`;
		expect(parser.parse(input)).toMatchSnapshot();
	});
	test("callbacks", () => {
		const input = `
A	"a" { callback_of_A(); }
B	/b/ { callback_of_B(); }

$S : T { callback_of_S(); };
T : A { callback_of_T_1(); } | E { callback_of_T_2(); } | { callback_of_T_3(); };
E : { callback_of_E(); } | B;
`;
		expect(parser.parse(input)).toMatchSnapshot();
	});
	test("ex-callbacks", () => {
		const input = `
#lex_end { lex_end_callback(); }
#lex_begin { lex_begin_callback(); }
#lex_default { lex_default_callback(); }
A	"a"

#default { grammar_default_callback(); }
$S : A;
`;
		expect(parser.parse(input)).toMatchSnapshot();
	});
	test("callback delimiters", () => {
		const input = `
A	"a" {{ if(1+1===3){ foo(); } }}

$S : T %{ const s = {}; }%;
T : E %%{ const t = "}%, }}%, }}%%, }%%%, }}%%%"; }%%;
E : { const e = "}%"+"}}"; } | A;
`;
		expect(parser.parse(input)).toMatchSnapshot();
	});
});
// TODO: languageファイルにコールバックも記述可能にして、それを読み取れるようにする
