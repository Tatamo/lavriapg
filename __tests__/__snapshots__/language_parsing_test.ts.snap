// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`syntax functions test callback delimiters 1`] = `
Object {
  "grammar": Object {
    "rules": Array [
      Object {
        "callback": " const s = {}; ",
        "ltoken": "S",
        "pattern": Array [
          "T",
        ],
      },
      Object {
        "callback": " const t = \\"}%, }}%, }}%%, }%%%, }}%%%\\"; ",
        "ltoken": "T",
        "pattern": Array [
          "E",
        ],
      },
      Object {
        "callback": " const e = \\"}%\\"+\\"}}\\"; ",
        "ltoken": "E",
        "pattern": Array [],
      },
      Object {
        "ltoken": "E",
        "pattern": Array [
          "A",
        ],
      },
    ],
    "start_symbol": "S",
  },
  "lex": Object {
    "rules": Array [
      Object {
        "callback": " if(1+1===3){ foo(); } ",
        "pattern": "a",
        "token": "A",
      },
    ],
  },
}
`;

exports[`syntax functions test callbacks 1`] = `
Object {
  "grammar": Object {
    "rules": Array [
      Object {
        "callback": " callback_of_S(); ",
        "ltoken": "S",
        "pattern": Array [
          "T",
        ],
      },
      Object {
        "callback": " callback_of_T_1(); ",
        "ltoken": "T",
        "pattern": Array [
          "A",
        ],
      },
      Object {
        "callback": " callback_of_T_2(); ",
        "ltoken": "T",
        "pattern": Array [
          "E",
        ],
      },
      Object {
        "callback": " callback_of_T_3(); ",
        "ltoken": "T",
        "pattern": Array [],
      },
      Object {
        "callback": " callback_of_E(); ",
        "ltoken": "E",
        "pattern": Array [],
      },
      Object {
        "ltoken": "E",
        "pattern": Array [
          "B",
        ],
      },
    ],
    "start_symbol": "S",
  },
  "lex": Object {
    "rules": Array [
      Object {
        "callback": " callback_of_A(); ",
        "pattern": "a",
        "token": "A",
      },
      Object {
        "callback": " callback_of_B(); ",
        "pattern": /b/,
        "token": "B",
      },
    ],
  },
}
`;

exports[`syntax functions test ex-callbacks 1`] = `
Object {
  "grammar": Object {
    "begin_callback": " grammar_begin_callback(); ",
    "default_callback": " grammar_default_callback(); ",
    "end_callback": " grammar_end_callback(); ",
    "rules": Array [
      Object {
        "ltoken": "S",
        "pattern": Array [
          "A",
        ],
      },
    ],
    "start_symbol": "S",
  },
  "lex": Object {
    "begin_callback": " lex_begin_callback(); ",
    "default_callback": " lex_default_callback(); ",
    "end_callback": " lex_end_callback(); ",
    "rules": Array [
      Object {
        "pattern": "a",
        "token": "A",
      },
    ],
  },
}
`;

exports[`syntax functions test lex-state 1`] = `
Object {
  "grammar": Object {
    "rules": Array [
      Object {
        "ltoken": "S",
        "pattern": Array [
          "A",
          "B2",
          "C",
        ],
      },
    ],
    "start_symbol": "S",
  },
  "lex": Object {
    "rules": Array [
      Object {
        "pattern": /a/,
        "token": "A",
      },
      Object {
        "pattern": /b/,
        "states": Array [
          "state1",
          "state2",
        ],
        "token": "B",
      },
      Object {
        "pattern": /b/,
        "states": Array [
          "default",
        ],
        "token": "B2",
      },
      Object {
        "pattern": /c/,
        "token": "C",
      },
    ],
    "start_state": "state1",
  },
}
`;
